File: Asignaturas/Tercer Año/PDOO/Examenes/SolucionParcialPracticas2/P5-java/src/irrgarten/FuzzyPlayer.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;

/**
 *
 * @author ismael
 */
/**
 * Clase FuzzyPlayer que extiende de Player.
 * Representa un jugador con comportamiento difuso.
 */
public class FuzzyPlayer extends Player {

    /**
     * Constructor que crea un FuzzyPlayer a partir de otro Player.
     * 
     * @param other El jugador del cual se copiarán las propiedades.
     */
    public FuzzyPlayer(Player other) {
        super(other);
    }

    /**
     * Método que determina el movimiento del jugador.
     * 
     * @param direction La dirección preferida de movimiento.
     * @param validMoves Lista de movimientos válidos.
     * @return La dirección final del movimiento.
     */
    @Override
    public Directions move(Directions direction, ArrayList<Directions> validMoves) {
        Directions preference = super.move(direction, validMoves);
        Directions d = Dice.nextStep(preference, validMoves, getIntelligence());
        return d;
    }

    /**
     * Método que calcula la intensidad del ataque del jugador.
     * 
     * @return La intensidad del ataque.
     */
    @Override
    public float attack() {
        return (Dice.intensity(getStrength()) + sumWeapons());
    }

    /**
     * Método que calcula la energía defensiva del jugador.
     * 
     * @return La energía defensiva.
     */
    @Override
    protected float defensiveEnergy() {
        return (Dice.intensity(getIntelligence()) + sumShields());
    }

    /**
     * Método que devuelve una representación en cadena del jugador.
     * 
     * @return Una cadena que representa al jugador.
     */
    @Override
    public String toString() {
        String out = "[Fuzzy]";
        return out + super.toString();
    }
}
    

File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_MOVIMIENTOS_PREDEFINIDOS/P5-Ruby/lib/irrgarten/fuzzy_player.rb
#encoding: UTF-8
require_relative 'player'
require_relative 'dice'

# Module containing classes related to the Irrgarten game.
module Irrgarten
    # Clase que representa a un FuzzyPlayer en el juego Irrgarten.
    # Hereda de Player y añade lógica difusa para el movimiento y el ataque.
    #
    # @!author Ismael Sallami Moreno
    class FuzzyPlayer < Player
        # Inicializa un nuevo FuzzyPlayer copiando otro jugador.
        #
        # @param other [Player] el jugador a copiar
        # @return [void]
        def initialize(other)
            copy(other)
        end

        # Determina el siguiente movimiento basado en lógica difusa.
        #
        # @param direction [Symbol] la dirección a moverse
        # @param validMoves [Array<Symbol>] los movimientos válidos disponibles
        # @return [Symbol] la dirección elegida
        def move(direction, validMoves)
            preference = super(direction, validMoves)
            Dice.nextStep(preference, validMoves, intelligence)
        end

        # Calcula la intensidad del ataque basado en la fuerza y las armas.
        #
        # @return [float] la intensidad del ataque
        def attack
            Dice.intensity(@strength) + sumWeapons
        end

        # Devuelve una representación en cadena del FuzzyPlayer.
        #
        # @return [String] la representación en cadena
        def to_s
            "[Fuzzy] #{super.to_s} "
        end

        protected

        # Calcula la energía defensiva basada en la inteligencia y los escudos.
        #
        # @return [float] la energía defensiva
        def defensive_energy
            Dice.intensity(@intelligence) + sumShields
        end
    end
end

